File: Sources/CS2TextureCombiner/Services/NormalMapNormalization.swift
import Foundation

enum NormalMapNormalization {
    static func normalize(_ raster: inout ImageRaster) {
        for pixel in 0..<(raster.width * raster.height) {
            let index = pixel * 4
            let normalized = normalizedComponents(
                red: raster.bytes[index],
                green: raster.bytes[index + 1],
                blue: raster.bytes[index + 2],
                maximum: UInt8.max
            )
            raster.bytes[index] = normalized.0
            raster.bytes[index + 1] = normalized.1
            raster.bytes[index + 2] = normalized.2
        }
    }

    static func normalizedComponents<T: BinaryInteger>(
        red: T,
        green: T,
        blue: T,
        maximum: T
    ) -> (T, T, T) {
        let maxValue = Double(maximum)
        let x = (Double(red) / maxValue) * 2.0 - 1.0
        let y = (Double(green) / maxValue) * 2.0 - 1.0
        let z = (Double(blue) / maxValue) * 2.0 - 1.0
        let length = sqrt(x * x + y * y + z * z)

        guard length > 0.000_001 else {
            return (T(maxValue * 0.5), T(maxValue * 0.5), maximum)
        }

        func encode(_ value: Double) -> T {
            let unit = max(-1.0, min(1.0, value / length))
            return T(((unit * 0.5 + 0.5) * maxValue).rounded())
        }
        return (encode(x), encode(y), encode(z))
    }
}

File: Windows/CS2Combiner.Core/NormalMapNormalization.cs
namespace CS2Combiner.Core;

public static class NormalMapNormalization
{
    public static void Normalize(ImageRaster raster)
    {
        for (var pixel = 0; pixel < raster.Width * raster.Height; pixel++)
        {
            var index = pixel * 4;
            var normalized = NormalizedComponents(
                raster.Bytes[index],
                raster.Bytes[index + 1],
                raster.Bytes[index + 2]);
            raster.Bytes[index] = normalized.Red;
            raster.Bytes[index + 1] = normalized.Green;
            raster.Bytes[index + 2] = normalized.Blue;
        }
    }

    public static (byte Red, byte Green, byte Blue) NormalizedComponents(
        byte red,
        byte green,
        byte blue)
    {
        const double maximum = byte.MaxValue;
        var x = red / maximum * 2.0 - 1.0;
        var y = green / maximum * 2.0 - 1.0;
        var z = blue / maximum * 2.0 - 1.0;
        var length = Math.Sqrt(x * x + y * y + z * z);

        if (length <= 0.000_001)
        {
            return ((byte)(maximum * 0.5), (byte)(maximum * 0.5), byte.MaxValue);
        }

        static byte Encode(double value, double vectorLength)
        {
            var unit = Math.Max(-1.0, Math.Min(1.0, value / vectorLength));
            return (byte)Math.Round(
                (unit * 0.5 + 0.5) * byte.MaxValue,
                MidpointRounding.AwayFromZero);
        }

        return (Encode(x, length), Encode(y, length), Encode(z, length));
    }
}
